src/main/java/crawlercommons/mimetypes/MimeTypeDetector.java

```diff
@@ -95,43 +104,65 @@ public MimeTypeDetector() {
 
         // Add all text patterns without and with a BOM.
-        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<?xml"));
-        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<?xml", true));
-        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<?XML"));
-        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<?XML", true));
-        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<!--"));
-        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<!--", true));
-
-        mimeTypes.add(new MimeTypeEntry(TEXT_MIMETYPES[0], "http://"));
-        mimeTypes.add(new MimeTypeEntry(TEXT_MIMETYPES[0], "http://", true));
-        mimeTypes.add(new MimeTypeEntry(TEXT_MIMETYPES[0], "https://"));
-        mimeTypes.add(new MimeTypeEntry(TEXT_MIMETYPES[0], "https://", true));
+        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<?xml", true, true));
+        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<?XML", true, true));
+        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<!--", true, true));
+        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<urlset", true, true));
+        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<sitemapindex", true, true));
+        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<rss", true, true));
+        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<feed", true, true));
+        mimeTypes.add(new MimeTypeEntry(XML_MIMETYPES[0], "<rdf", true, true));
+
+        mimeTypes.add(new MimeTypeEntry(TEXT_MIMETYPES[0], "http://", true, true));
+        mimeTypes.add(new MimeTypeEntry(TEXT_MIMETYPES[0], "https://", true, true));
 
         mimeTypes.add(new MimeTypeEntry(GZIP_MIMETYPES[0], "\037\213"));
         mimeTypes.add(new MimeTypeEntry(GZIP_MIMETYPES[0], 0x1F, 0x8B));
 
         maxPatternLength = 0;
         for (MimeTypeEntry entry : mimeTypes) {
-            maxPatternLength = Math.max(maxPatternLength, entry.getPattern().length);
+            int length = entry.getPattern().length;
+            if (entry.allowBOM)
+                length += UTF8_BOM.length;
+            if (entry.allowLeadingSpace)
+                length += LEADING_WHITESPACE_MAX_SKIP;
+            maxPatternLength = Math.max(maxPatternLength, length);
         }
     }
 
     public String detect(byte[] content) {
-        for (MimeTypeEntry entry : mimeTypes) {
-            if (patternMatches(entry.getPattern(), content, 0, content.length)) {
-                return entry.getMimeType();
-            }
-        }
-
-        // No mime-type detected.
-        return null;
+        return detect(content, 0, content.length);
     }
 
     public String detect(byte[] content, int offset, int length) {
+        int offsetBOM = -1;
+        int offsetSpace = -1;
         for (MimeTypeEntry entry : mimeTypes) {
             if (patternMatches(entry.getPattern(), content, offset, length)) {
                 return entry.getMimeType();
             }
+            if (entry.allowBOM) {
```

Comment:
BOM has to be at the beginning. I also think we should simplify and assume offset == 0 always (nobody calls detect with non-zero offset). So just have an isBOM check (which advances the offset by BOM length), and then an isWhitespace check (same thing) and then the previous logic.

Reply:
> BOM has to be at the beginning.

Of course, but in my test set there are sitemaps with two BOMs. We would need also to change the BOMInputStream (also to skip white space), but detection comes first.

> assume offset == 0 always (nobody calls detect with non-zero offset)

Agreed.